src/hdmf/container.py

```diff
@@ -751,14 +754,29 @@ def _generate_field_html(self, key, value, level, access_code):
 
 
     def _generate_array_html(self, array, level):
-        """Generates HTML for array data"""
+        """Generates HTML for array data (e.g., NumPy arrays, HDF5 datasets, Zarr datasets and DataIO objects)."""
 
-        read_io = self.get_read_io()  # if the Container was read from file, get IO object
-        if read_io is not None: # Note that sometimes numpy array have a read_io attribute
-            repr_html = read_io.generate_dataset_html(array)
-        else:
+        is_numpy_array = isinstance(array, np.ndarray)
+        it_was_read_with_io = self.get_read_io() is not None
+        is_data_io = isinstance(array, DataIO)
+
+        if is_numpy_array:
             array_info_dict = get_basic_array_info(array)
             repr_html = generate_array_html_repr(array_info_dict, array, "NumPy array")
+        elif is_data_io:
+            array_info_dict = get_basic_array_info(array.data)
+            repr_html = generate_array_html_repr(array_info_dict, array.data, "DataIO")
+        elif it_was_read_with_io:
+            # The backend handles the representation here. Two special cases worth noting:
+            # 1. Array-type attributes (e.g., start_frame in ImageSeries) remain NumPy arrays
+            #    even when their parent container has an IO
+            # 2. Data may have been modified after being read from storage
+            read_io = self.get_read_io()
+            repr_html = read_io.generate_dataset_html(array)
+        else:  # Not sure which object could get here
```

Comment:
Based on our discussion above, I believe any `AbstractDataChunkIteator` would currently go to the else statement. Also, if the data was specified as a `list` or `tuple` , I think it would end up in `else`

Reply:
Those objects will fail to enter this function:

https://github.com/h-mayorquin/hdmf/blob/4a0dbd9b635add255e49803be5d03d0e6600d89c/src/hdmf/container.py#L730-L733

Reply:
Thanks for the clarification. I did not see that there was additional logic outside of the function for this.